libs/grout/README.md

```diff
@@ -0,0 +1,160 @@
+# Grout
+
+A library to create type-safe RPC glue code.
+
+Grout lets you define a server API as a plain TS object with async methods.
+Based on this API definition, Grout will create Express route handlers on the
+server side and a type-safe client on the client side, abstracting away the
+details of the underlying HTTP requests.
+
+## Installation
+
+Grout must be installed as a dependency in both your client and server projects.
+
+## Usage
+
+### Server
+
+```ts
+import * as grout from '@votingworks/grout';
+
+interface Person {
+  name: string;
+  age: number;
+}
+
+// Create an API definition by building an object with async methods.
+const api = grout.createApi({
+  // Each method should return a Promise. The data returned should be simple TS
+  // values - not fancy class instances. Values that can be easily serialized to
+  // JSON. More details on this below.
+  async getAllPeople(): Promise<Person[]> {
+    return store.getAllPeople(); // Assume this accesses the database, etc.
+  },
+
+  // Methods can take input, but it must be packaged into a single object.
+  // Think of it like using named parameters, or React component props.
+  async getPersonByName(input: { name: string }): Promise<Person | undefined> {
+    return store.getPersonByName(input.name);
+  },
+
+  // For known errors (e.g. invalid input), use a Result type so that the client
+  // forced to handle them explicitly.
+  async updatePersonAge(input: {
+    name: string;
+    age: number;
+  }): Promise<Result<void, Error>> {
+    if (age <= 0) {
+      return err(new Error('Age must be at least 1.'));
+    }
+    await store.updatePersonAge(input.name, input.age);
+    return ok();
+  },
+});
+
+const app = express();
+
+// When registering the API with an Express app, you must use a separate
+// Router. That way you can isolate the Grout API routes from other routes in
+// your app.
+const router = express.Router();
+
+// You must have the router use the `express.text()` middleware with type
+// `application/json` exactly as shown below.
+router.use(express.text({ type: 'application/json' }));
+
+// Then, simply register the API routes.
+grout.registerRoutes(router, api);
+
+// Finally, mount the router on your app. You can use any path prefix you want -
+// here we use `/api`. It's important that you don't use any other body-parsing
+// middleware upstream of the router - e.g. don't call `app.use(express.json())`
+// before this.
+app.use('/api', router);
+
+// Don't forget to export the API type - we'll need it for the client.
+export type MyApi = typeof api;
+```
+
+### Client
+
+```ts
+import * as grout from '@votingworks/grout';
+
+// First, we need to import the API type from the server. This will *not* import
+// any of the actual server code - just the type definition.
+import type { MyApi } from './server';
```

Comment:
Thanks for all the sample code and comments in this README! Very helpful format!

Reply:
How do you envision these imports actually working? Would the type just live in `api` and be imported by both the server and client?

Reply:
Here's an example that I think I got working in VxScan:
- Exporting the type from the backend: https://github.com/votingworks/vxsuite/pull/2854/files#diff-041bc090c8c2462dccf83998ef49612c01e865111f24d78b8cda709aabe6a6abR60
- Importing the type on the frontend: https://github.com/votingworks/vxsuite/pull/2854/files#diff-c10dee518b69cf839da28bbee8f6e170d9717f29ea1898cf5ec2c4388f3b7463R2